pkg/api/api.go

```diff
@@ -43,97 +43,138 @@ type podInfo struct {
-	a := &API{
-		log:                  log,
-		readiness:            false,
-		currentConfiguration: currentConfiguration,
-		apiPort:              apiPort,
-		apiHost:              apiHost,
-		podLister:            podLister,
-		meshNamespace:        meshNamespace,
+func NewAPI(log logrus.FieldLogger, apiPort int32, apiHost string, client kubernetes.Interface, meshNamespace string) (*API, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{"component": "maesh-mesh"},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create label selector: %w", err)
+	}
+
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(client, k8s.ResyncPeriod,
+		informers.WithNamespace(meshNamespace),
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = selector.String()
+		}))
+
+	podLister := informerFactory.Core().V1().Pods().Lister()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	informerFactory.Start(ctx.Done())
+
+	for t, ok := range informerFactory.WaitForCacheSync(ctx.Done()) {
+		if !ok {
+			return nil, fmt.Errorf("timed out while waiting for informer cache to sync: %s", t)
+		}
 	}
 
-	if err := a.Init(); err != nil {
-		log.Errorln("Could not initialize API")
+	router := mux.NewRouter()
+
+	a := &API{
+		Server: http.Server{
+			Addr:         fmt.Sprintf("%s:%d", apiHost, apiPort),
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 5 * time.Second,
+			Handler:      router,
+		},
+		configuration: safe.New(provider.NewDefaultDynamicConfig()),
+		topology:      safe.New(topology.NewTopology()),
+		readiness:     safe.New(false),
+		podLister:     podLister,
+		meshNamespace: meshNamespace,
+		log:           log,
 	}
 
-	return a
+	router.HandleFunc("/api/configuration/current", a.getCurrentConfiguration)
+	router.HandleFunc("/api/topology/current", a.getCurrentTopology)
+	router.HandleFunc("/api/status/nodes", a.getMeshNodes)
+	router.HandleFunc("/api/status/node/{node}/configuration", a.getMeshNodeConfiguration)
+	router.HandleFunc("/api/status/readiness", a.getReadiness)
+
+	return a, nil
 }
 
-// Init handles any api initialization.
-func (a *API) Init() error {
-	a.log.Debugln("Initializing API")
+// SetReadiness sets the readiness flag in the API.
+func (a *API) SetReadiness(isReady bool) {
+	readiness, ok := a.readiness.Get().(bool)
 
-	a.router = mux.NewRouter()
+	if ok && readiness == isReady {
+		return
+	}
 
-	a.router.HandleFunc("/api/configuration/current", a.getCurrentConfiguration)
-	a.router.HandleFunc("/api/status/nodes", a.getMeshNodes)
-	a.router.HandleFunc("/api/status/node/{node}/configuration", a.getMeshNodeConfiguration)
-	a.router.HandleFunc("/api/status/readiness", a.getReadiness)
+	a.readiness.Set(isReady)
 
-	return nil
+	if isReady {
```

Comment:
I don't think that this test is needed, maybe it's enough to display the ready value in the debug message?